File: contents/convolutions/1d/code/csharp/1DConvolution.cs
using System;
using System.IO;

namespace Convolution1D
{
    public class Convolution1D
    {
        // Creates a sawtooth function with the given length.
        static double[] CreateSawtooth(int length)
        {
            var array = new double[length];
            for (var i = 0; i < length; i++)
                array[i] = (i + 1) / 200f;
            return array;
        }

        // Normalizes the given array.
        static void Normalize(double[] array)
        {
            var norm = Norm(array);
            for (var i = 0; i < array.Length; i++)
                array[i] /= norm;
        }

        // Calculates the norm of the array.
        static double Norm(double[] array)
        {
            var sum = 0.0;
            for (var i = 0; i < array.Length; i++)
                sum += Math.Pow(array[i], 2);
            return Math.Sqrt(sum);
        }

        // Modulus function which handles negative values properly.
        // Assumes that y >= 0.
        static int Mod(int x, int y) => ((x % y) + y) % y;

        static double[] ConvolveCyclic(double[] signal, double[] filter)
        {
            var outputSize = Math.Max(signal.Length, filter.Length);

            // Convolutional output.
            var output = new double[outputSize];
            var sum = 0.0;

            for (var i = 0; i < outputSize; i++)
            {
                for (var j = 0; j < outputSize; j++)
                {
                    if (Mod(i - j, outputSize) < filter.Length)
                    {
                        sum += signal[Mod(j - 1, outputSize)] * filter[Mod(i - j, outputSize)];
                    }
                }

                output[i] = sum;
                sum = 0.0;
            }

            return output;
        }

        static double[] ConvolveLinear(double[] signal, double[] filter, int outputSize)
        {
            // Convolutional output.
            var output = new double[outputSize];
            var sum = 0.0;

            for (var i = 0; i < outputSize; i++)
            {
                for (var j = Math.Max(0, i - filter.Length); j <= i; j++)
                {
                    if (j < signal.Length && (i - j) < filter.Length)
                    {
                        sum += signal[j] * filter[i - j];
                    }
                }

                output[i] = sum;
                sum = 0.0;
            }

            return output;
        }

        static void Main()
        {
            // Create sawtooth functions for x and y.
            var x = CreateSawtooth(200);
            var y = CreateSawtooth(200);

            // Normalization is not strictly necessary, but good practice.
            Normalize(x);
            Normalize(y);

            // Full convolution, output will be the size of x + y - 1.
            var fullLinearOutput = ConvolveLinear(x, y, x.Length + y.Length - 1);
            // Simple boundaries.
            var simpleLinearOutput = ConvolveLinear(x, y, x.Length);
            // Cyclic convolution.
            var cyclicOutput = ConvolveCyclic(x, y);

            // Output convolutions to different files for plotting.
            File.WriteAllText("full_linear.dat", String.Join(Environment.NewLine, fullLinearOutput));
            File.WriteAllText("simple_linear.dat", String.Join(Environment.NewLine, simpleLinearOutput));
            File.WriteAllText("cyclic.dat", String.Join(Environment.NewLine, cyclicOutput));
        }
    }
}

File: contents/convolutions/1d/code/rust/1d_convolution.rs
use std::fs::File;
use std::io::Write;
use std::cmp::max;

fn main() {
    let x = normalize(create_sawtooth(200));
    let y = normalize(create_sawtooth(200));

    let len_x = x.len();
    let len_y = y.len();

    let full_linear_output = convolve_linear(&x, &y, len_x + len_y - 1);
    let simple_linear_output = convolve_linear(&x, &y, len_x);
    let cyclic_output = convolve_cyclic(&x, &y);

    // Save the convolutions to plot them.
    // The way I do it is a little weird but it is to store the data the same way as the other programs
    let mut full_file = File::create("full_linear.dat").unwrap();
    for i in full_linear_output {
        writeln!(full_file, "{i}").unwrap();
    }
    full_file.sync_all().unwrap();

    let mut simple_file = File::create("simple_linear.dat").unwrap();
    for i in simple_linear_output {
        writeln!(simple_file, "{i}").unwrap();
    }
    simple_file.sync_all().unwrap();

    let mut cyclic_file = File::create("cyclic.dat").unwrap();
    for i in cyclic_output {
        writeln!(cyclic_file, "{i}").unwrap();
    }
    cyclic_file.sync_all().unwrap();
}

// Generates a sawtooth function with a given length.
fn create_sawtooth(length: usize) -> Vec<f64> {
    let mut array: Vec<f64> = Vec::with_capacity(length);
    for i in 0..length {
        array.push((i+1) as f64 / length as f64); // divide by length for normalization
    }
    array
}

// Normalizes the given array.
fn normalize(array: Vec<f64>) -> Vec<f64> {
    let norm = norm(&array);
    let mut output: Vec<f64> = Vec::with_capacity(array.len());

    for value in array {
        output.push(value / norm);
    }

    output
}

// Calculates the norm of an array.
fn norm(array: &[f64]) -> f64 {
    let sum = array.iter().map(|i| i * i).sum::<f64>();
    sum.sqrt()
}


// Modulus function that handles negative values correctly.
// Assumes that y >= 0.
fn modulus(x: isize, y: usize) -> usize {((x%y as isize) as usize + y) % y}

fn convolve_linear(signal: &Vec<f64>, filter: &Vec<f64>, output_size: usize) -> Vec<f64> {
    let mut output = Vec::with_capacity(output_size);

    for i in 0..(output_size as isize) {
        let mut sum: f64 = 0.;
        for j in max(0, i - filter.len() as isize)..=i {
            if j < signal.len() as isize && (i - j) < filter.len() as isize {
                sum += signal[j as usize] * filter[(i-j) as usize];
            }
        }
        output.push(sum);
    }

    output
}

fn convolve_cyclic(signal: &Vec<f64>, filter: &Vec<f64>) -> Vec<f64> {
    let output_size = max(signal.len(), filter.len());

    let mut output: Vec<f64> = Vec::with_capacity(output_size);
    for i in 0..output_size {
        let mut sum: f64 = 0.;
        for j in 0..output_size {
            if modulus((i - j) as isize, output_size) < filter.len() {
                sum += signal[modulus((j - 1) as isize, output_size)] * filter[modulus((i - j) as isize, output_size)];
            }
        }
        output.push(sum);
    }
    output
}